pennylane/transforms/op_transforms.py

```diff
@@ -0,0 +1,481 @@
+# limitations under the License.
+"""
+This module contains the @op_transform decorator.
+"""
+# pylint: disable=protected-access
+import functools
+import inspect
+import os
+import warnings
+
+import pennylane as qml
+
+
+class OperationTransformError(Exception):
+    """Raised when there is an error with op_transform logic"""
+
+
+def _make_tape(obj, wire_order, *args, **kwargs):
+    """Given an input object, which may be:
+
+    - an object such as a tape or a operation, or
+    - a callable such as a QNode or a quantum function
+      (alongside the callable arguments ``args`` and ``kwargs``),
+
+    this function constructs and returns the tape/operation
+    represented by the object.
+
+    The ``wire_order`` argument determines whether a custom wire ordering
+    should be used. It not provided, the wire ordering defaults to the
+    objects wire ordering accessed via ``obj.wires``.
+    """
+    if isinstance(obj, qml.QNode):
+        # user passed a QNode, get the tape
+        obj.construct(args, kwargs)
+        tape = obj.qtape
+        wires = obj.device.wires
+
+    elif isinstance(obj, qml.tape.QuantumTape):
+        # user passed a tape
+        tape = obj
+        wires = tape.wires
+
+    elif inspect.isclass(obj) and issubclass(obj, qml.operation.Operator):
+        with qml.tape.stop_recording():
+            tape = obj(*args, **kwargs)
+
+        wires = tape.wires
+
+    elif callable(obj):
+        # user passed something that is callable but not a tape or QNode.
+        tape = qml.transforms.make_tape(obj)(*args, **kwargs)
+        wires = tape.wires
+
+        # raise exception if it is not a quantum function
+        if len(tape.operations) == 0 and len(tape.measurements) == 0:
+            raise OperationTransformError("Quantum function contains no quantum operations")
+
+    # if no wire ordering is specified, take wire list from tape/device
+    wire_order = wires if wire_order is None else qml.wires.Wires(wire_order)
+
+    # check that all wire labels in the circuit are contained in wire_order
+    if not set(tape.wires).issubset(wire_order):
+        raise OperationTransformError(
+            f"Wires in circuit {tape.wires.tolist()} are inconsistent with "
+            f"those in wire_order {wire_order.tolist()}"
+        )
+
+    return tape, wire_order
+
+
+class op_transform:
+    r"""Class for registering an operator transform that takes one or more operators,
+    and returns a classical representation.
+
+    Using ``op_transform`` is not necessary in most cases; simply define a
+    standard Python function that accepts an operator and returns the
+    computed quantity.
+
+    However, this registration class is useful if you wish to easily create
+    a function that:
+
+    - Supports datastructures that may contain multiple operations, such as
+      a tape, QNode, or qfunc.
+
+    - Supports being used with a functional transform UI.

```

Comment:
Reading this again, I feel that the odd feeling that came up with the syntax is because this class does two things instead of one. These two things are well explained here.

Could we have two different objects for:
* Defining an operation transform;
* Registering transforms to the same name such that they work on different data structures like an operation, tape, QNode, or qfunc?

Reply:
@antalszava I completely agree! @mariaschuld also suggested we have a separate

```python
qml.add_transform(existing_transform, new_func)
```

However, I think this is a rare case where I am not happy 100% with the existing design, _but_ I think there needs a dedicated design/prototyping period that we don't yet have time for.

My plan is:

- we have this decorator in for this release just so that we have the functionality available
- at the same time, we do not advertise this new transform/API.
- after this release, I would like to go back to the drawing board and see if we can design a _singular_ `@qml.transform` decorator that smartly chooses what underlying type of transform to use. And maybe also handles registration?

Reply:
What do you think of this plan?

Reply:
> What would we like to use the functionality for?

Mostly the ability to quickly add the correct UI in a way that allows us to avoid re-using the same boilerplate code over and over. This will help with several operator functions that are needed for the operator refactor before this release:

- https://github.com/PennyLaneAI/pennylane/pull/2256
- https://github.com/PennyLaneAI/pennylane/pull/2248
- https://github.com/PennyLaneAI/pennylane/pull/2241

> One worry I have is that we'd be building on functionality that we know will refactor soon, resulting in more efforts required overall.

I think this is inevitable, we need to refactor and re-align _all_ the transforms :) including batch and qfunc. Currently, most of the code in this class is spread and duplicated across the codebase, here it is collected and tested independently.